src/eos-updater-poll.c

```diff
@@ -257,6 +257,247 @@ async_result_cb (GObject      *source_object,
+                                               GVariant            **out_commit,
+                                               GCancellable         *cancellable,
+                                               GError              **error)
+{
+  g_autofree gchar *upgrade_refspec = NULL;
+  g_autofree gchar *remote = NULL;
+  g_autofree gchar *ref = NULL;
+  g_autoptr(OstreeCollectionRef) collection_ref = NULL;
+  g_autofree gchar *new_refspec = NULL;
+  g_autofree gchar *version = NULL;
+  g_autofree gchar *checksum = NULL;
+  g_autoptr(GVariant) commit = NULL;
+
+  g_return_val_if_fail (out_refspec != NULL, FALSE);
+  g_return_val_if_fail (out_remote != NULL, FALSE);
+  g_return_val_if_fail (out_ref != NULL, FALSE);
+  g_return_val_if_fail (out_collection_ref != NULL, FALSE);
+  g_return_val_if_fail (out_new_refspec != NULL, FALSE);
+  g_return_val_if_fail (out_checksum != NULL, FALSE);
+  g_return_val_if_fail (out_version != NULL, FALSE);
+  g_return_val_if_fail (out_commit != NULL, FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  if (!get_refspec_to_upgrade_on (&upgrade_refspec, NULL, NULL, NULL, error))
+    return FALSE;
+
+  if (!fetch_latest_commit (repo,
+                            cancellable,
+                            upgrade_refspec,
+                            NULL,
+                            &checksum,
+                            &new_refspec,
+                            &version,
+                            error))
+    return FALSE;
+
+  *out_refspec = g_steal_pointer (&upgrade_refspec);
+  *out_remote = g_steal_pointer (&remote);
+  *out_ref = g_steal_pointer (&ref);
+  *out_collection_ref = g_steal_pointer (&collection_ref);
+  *out_new_refspec = g_steal_pointer (&new_refspec);
+  *out_checksum = g_steal_pointer (&checksum);
+  *out_version = g_steal_pointer (&version);
+  *out_commit = g_steal_pointer (&commit);
+
+  return TRUE;
+}
+
+static gboolean
+check_for_update_following_checkpoint_if_allowed (OstreeRepo           *repo,
+                                                  gchar               **out_refspec_to_upgrade_on,
+                                                  gchar               **out_remote_to_upgrade_on,
+                                                  gchar               **out_ref_to_upgrade_on,
+                                                  OstreeCollectionRef **out_collection_ref_to_upgrade_on,
+                                                  gchar               **out_redirected_refspec_to_upgrade_on,
+                                                  gchar               **out_checksum,
+                                                  gchar               **out_version,
+                                                  GVariant            **out_commit,
+                                                  GCancellable         *cancellable,
+                                                  GError              **error)
```

Comment:
Having a function with 11 arguments like this is not ideal :( Ideally what I wanted to be able to do here was have an abstraction that allows callers looking for which commits and branches to follow to do the right thing in a single function call ... without then having to go and fetch all the commit information again later on.

Previously the control flow here was a little more straightforward - we would work out which refspec to pull on, then pull that refspec and work out whether an update was available. Now we need to always pull the booted refspec in order to work out whether or not to even unconditionally use the booted refspec, then figure out if we need to use the checkpoint refspec if available, then pull commits from the right refspec.

Ideally I'd like to avoid pulling commits twice, hence the preservation of all the outparams...

Reply:
You could put the out-params into a structure, allocate that on the stack in the caller, and pass that in to this function by reference. That would reduce the parameter explosion fairly trivially.